airbyte-integrations/connectors/destination-pubsub/src/main/java/io/airbyte/integrations/destination/pubsub/PubsubConsumer.java

```diff
@@ -0,0 +1,138 @@
+import io.airbyte.integrations.base.JavaBaseConstants;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.AirbyteMessage.Type;
+import io.airbyte.protocol.models.AirbyteRecordMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import java.io.ByteArrayInputStream;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PubsubConsumer extends FailureTrackingAirbyteMessageConsumer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PubsubConsumer.class);
+  private final JsonNode config;
+  private final ConfiguredAirbyteCatalog catalog;
+  private final Consumer<AirbyteMessage> outputRecordCollector;
+  private final Map<AirbyteStreamNameNamespacePair, Map<String, String>> attributes;
+  private Publisher publisher;
+  private AirbyteMessage lastStateMessage;
+
+  public PubsubConsumer(JsonNode config,
+      ConfiguredAirbyteCatalog catalog,
+      Consumer<AirbyteMessage> outputRecordCollector) {
+    this.outputRecordCollector = outputRecordCollector;
+    this.config = config;
+    this.catalog = catalog;
+    this.lastStateMessage = null;
+    this.attributes = Maps.newHashMap();
+    this.publisher = null;
+    LOGGER.info("initializing consumer.");
+  }
+
+  @Override
+  protected void startTracked() throws Exception {
+    // get publisher
+    final String projectId = config.get(PubsubDestination.CONFIG_PROJECT_ID).asText();
+    final String topicName = config.get(PubsubDestination.CONFIG_TOPIC_ID).asText();
+    TopicName topic = TopicName.of(projectId, topicName);
+    final String credentialsString =
+        config.get(PubsubDestination.CONFIG_CREDS).isObject() ? Jsons.serialize(config.get(
+            PubsubDestination.CONFIG_CREDS))
+            : config.get(PubsubDestination.CONFIG_CREDS).asText();
+    final ServiceAccountCredentials credentials = ServiceAccountCredentials
+        .fromStream(new ByteArrayInputStream(credentialsString.getBytes(Charsets.UTF_8)));
+    publisher = Publisher.newBuilder(topic)
+        .setEnableMessageOrdering(true)
+        .setCredentialsProvider(FixedCredentialsProvider.create(credentials)).build();
+    for (final ConfiguredAirbyteStream configStream : catalog.getStreams()) {
+      final Map<String, String> attrs = Maps.newHashMap();
+      var key = AirbyteStreamNameNamespacePair.fromAirbyteSteam(configStream.getStream());
+      attrs.put(PubsubDestination.STREAM, key.getName());
+      if (!Strings.isNullOrEmpty(key.getNamespace())) {
+        attrs.put(PubsubDestination.NAMESPACE, key.getNamespace());
+      }
+      attributes.put(key, attrs);
+    }
+  }
+
+  @Override
+  protected void acceptTracked(AirbyteMessage msg) throws Exception {
+    if (msg.getType() == Type.STATE) {
+      lastStateMessage = msg;
```

Comment:
we can also run `outputRecordCollector.accept(lastStateMessage)` here to indicate that all previous messages have been written. This is good in the case of halfway failures, since it would allow us to resume a sync from the most recent state message rather from the beginning (which is currently the behavior)

Reply:
That makes sense, will change it